lms/static/scripts/frontend_apps/utils/fetch.js

```diff
@@ -0,0 +1,87 @@
+import { useEffect, useState } from 'preact/hooks';
+
+/**
+ * @typedef FetchIncomplete
+ * @prop {'none'|'loading'} state
+ */
+
+/**
+ * @template T
+ * @typedef FetchOk
+ * @prop {'ok'} state
+ * @prop {T} data
+ */
+
+/**
+ * @typedef FetchError
+ * @prop {'error'} state
+ * @prop {Error} error
+ */
+
+/**
+ * @template T
+ * @typedef {FetchIncomplete|FetchOk<T>|FetchError} FetchResult
+ */
+
+/**
+ * Function that fetches data (eg. from the API). The passed signal will be
+ * aborted if the fetched data is no longer needed (eg. due to a component being
+ * unmounted).
+ *
+ * @template T
+ * @typedef {(signal: AbortSignal) => Promise<T>} Fetcher
+ */
+
+/**
+ * Hook that fetches data from the backend API or some other async data source.
+ *
+ * This hook handles:
+ *
+ *  - Performing the initial fetch, and then re-fetching if the data to fetch
+ *    changes (indicated by the `fetcher` callback changing)
+ *  - Maintaining the state of the fetch (idle, fetching, fetched, error) and
```

Comment:
It'd be nice if these prose state descriptions were more closely matched to the defined `state` strings.